ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForMinor.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hive.common.util.HiveStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Builds query strings that help with query-based MINOR compaction of CRUD.
+ */
+class CompactionQueryBuilderForMinor extends CompactionQueryBuilder {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CompactionQueryBuilderForMinor.class.getName());
+
+  /**
+   * Construct a CompactionQueryBuilderMinor with required params.
+   *
+   * @param compactionType major or minor or rebalance, e.g. CompactionType.MAJOR.
+   *                       Cannot be null.
+   * @param operation query's Operation e.g. Operation.CREATE.
+   * @throws IllegalArgumentException if compactionType is null
+   */
+  CompactionQueryBuilderForMinor(CompactionType compactionType, Operation operation, String resultTableName) {
+    super(compactionType, operation, false, resultTableName);
+  }
+
+  protected void buildSelectClauseForInsert(StringBuilder query) {
+    // Need list of columns for major crud, mmmajor partitioned, mmminor
+    query.append("`operation`, `originalTransaction`, `bucket`, `rowId`, `currentTransaction`, `row`");
+  }
+
+  protected void getSourceForInsert(StringBuilder query) {
+    if (sourceTabForInsert != null) {
+      query.append(sourceTabForInsert);
+    } else {
+      query.append(sourceTab.getDbName()).append(".").append(sourceTab.getTableName());
+    }
+    query.append(" ");
+  }
+
+  protected void buildWhereClauseForInsert(StringBuilder query) {
+    if (validWriteIdList != null) {
+      if (validWriteIdList != null) {
+        long[] invalidWriteIds = validWriteIdList.getInvalidWriteIds();
+        if (invalidWriteIds.length > 0) {
+          query.append(" where `originalTransaction` not in (")
+              .append(StringUtils.join(ArrayUtils.toObject(invalidWriteIds), ",")).append(")");
+        }
+      }
+    }
+  }
+
+  protected void getDdlForCreate(StringBuilder query) {
+    defineColumns(query);
+
+    // PARTITIONED BY. Used for parts of minor compaction.
+    if (isPartitioned) {
+      query.append(" PARTITIONED BY (`file_name` STRING) ");
+    }
+
+    // CLUSTERED BY. (bucketing)
+    int bucketingVersion = 0;
+    if (CompactionType.MINOR.equals(compactionType)) {
+      bucketingVersion = getMinorCrudBucketing(query, bucketingVersion);
+    }
+
+    // STORED AS / ROW FORMAT SERDE + INPUTFORMAT + OUTPUTFORMAT
+    query.append(" stored as orc");
+
+    // LOCATION
+    if (location != null) {
+      query.append(" LOCATION '").append(HiveStringUtils.escapeHiveCommand(location)).append("'");
+    }
+
+    // TBLPROPERTIES
+    addTblProperties(query, bucketingVersion);
+  }
+
+  /**
+   * Define columns of the create query.
+   */
+  private void defineColumns(StringBuilder query) {
+    if (sourceTab == null) {
```

Comment:
Why are doing this null & return here and even other places can't we wrap our logic within if(sourceTab!=null)